typing/rec_check.ml

```diff
@@ -650,8 +976,9 @@ and is_destructuring_pattern : Typedtree.pattern -> bool =
     | Tpat_exception _ -> false
 
 let is_valid_recursive_expression idlist expr =
-  let ty = expression (build_unguarded_env idlist) expr in
-  match Use.unguarded ty, Use.dependent ty, classify_expression expr with
+  let ty = expression expr Return in
+  match Env.unguarded ty idlist, Env.dependent ty idlist,
+        classify_expression expr with
   | _ :: _, _, _ (* The expression inspects rec-bound variables *)
   | _, _ :: _, Dynamic -> (* The expression depends on rec-bound variables
```

Comment:
The first underscore can be turned into a `[]` which saves a few brain cycles to the reader.